File: uigesturerecognizer/src/main/java/it/sephiroth/android/library/uigestures/UILongPressGestureRecognizer.java
package it.sephiroth.android.library.uigestures;

import android.content.Context;
import android.graphics.PointF;
import android.os.Message;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;
import android.view.MotionEvent;
import android.view.ViewConfiguration;

/**
 * Created by alessandro crugnola on 11/20/16.
 * UIGestureRecognizer
 */
@SuppressWarnings ("unused")
public class UILongPressGestureRecognizer extends UIGestureRecognizer implements UIContinuousRecognizer {
    private long longPressTimeOut = TAP_TIMEOUT + LONG_PRESS_TIMEOUT;

    // request to change the current state to Failed
    private static final int MESSAGE_FAILED = 1;
    // request to change the current state to Possible
    private static final int MESSAGE_RESET = 2;
    // we handle the action_pointer_up received in the onTouchEvent with a delay
    // in order to check how many fingers were actually down when we're checking them
    // in the action_up.
    private static final int MESSAGE_POINTER_UP = 3;
    // post handle the long press event
    private static final int MESSAGE_LONG_PRESS = 4;

    // number of required fingers (default is 1)
    private int mTouchesRequired = 1;
    // number of required taps (default is 1)
    private int mTapsRequired = 0;

    private boolean mAlwaysInTapRegion;
    private float mDownFocusX;
    private float mDownFocusY;
    private float mTouchSlopSquare;
    private float mAllowableMovementSquare;
    private boolean mStarted;

    private int mNumTaps = 0;
    private int mNumTouches = 0;
    private final PointF mCurrentLocation;
    private boolean mFireEvents;
    private boolean mBegan;

    /**
     * UILongPressGestureRecognizer looks for long-press gestures.
     * The user must press one or more fingers on a view and hold them there for a minimum period of time before the action
     * triggers.
     * While down, the user’s fingers may not move more than a specified distance; if they move beyond the specified distance,
     * the gesture fails.
     */
    public UILongPressGestureRecognizer(@Nullable final Context context) {
        super(context);

        mStarted = false;
        mBegan = false;
        mCurrentLocation = new PointF();

        int touchSlop;

        if (context == null) {
            //noinspection deprecation
            touchSlop = ViewConfiguration.getTouchSlop();
        } else {
            final ViewConfiguration configuration = ViewConfiguration.get(context);
            touchSlop = configuration.getScaledTouchSlop();
        }
        mTouchSlopSquare = touchSlop * touchSlop;
        mAllowableMovementSquare = mTouchSlopSquare;
    }

    @Override
    protected void handleMessage(final Message msg) {
        switch (msg.what) {
            case MESSAGE_RESET:
                logMessage(Log.INFO, "handleMessage(MESSAGE_RESET)");
                handleReset();
                break;

            case MESSAGE_FAILED:
                logMessage(Log.INFO, "handleMessage(MESSAGE_FAILED)");
                handleFailed();
                break;

            case MESSAGE_POINTER_UP:
                logMessage(Log.INFO, "handleMessage(MESSAGE_POINTER_UP)");
                mNumTouches = msg.arg1;
                break;

            case MESSAGE_LONG_PRESS:
                logMessage(Log.INFO, "handleMessage(MESSAGE_LONG_PRESS)");
                handleLongPress();
                break;

            default:
                break;
        }
    }

    @Override
    public int getNumberOfTouches() {
        return mNumTouches;
    }

    @Override
    public void onStateChanged(@NonNull final UIGestureRecognizer recognizer) {
        logMessage(Log.VERBOSE, "onStateChanged(%s, %s)", recognizer, recognizer.getState());
        logMessage(Log.VERBOSE, "started: %b, state: %s", mStarted, getState());

        if (recognizer.getState() == State.Failed && getState() == State.Began) {
            mFireEvents = true;
            stopListenForOtherStateChanges();
            fireActionEvent();

            if (mBegan) {
                setState(State.Changed);
            }

        } else if (recognizer.inState(State.Began, State.Ended) && mStarted && getState() == State.Possible) {
            stopListenForOtherStateChanges();
            removeMessages();
            mStarted = false;
            setState(State.Failed);
        }
    }

    /**
     * @param value The number of required taps for this recognizer to succeed.<br />
     *              Default value is 1
     * @since 1.0.0
     */
    public void setNumberOfTapsRequired(int value) {
        mTapsRequired = value;
    }

    /**
     * @param value The number of required touches for this recognizer to succeed.<br />
     *              Default value is 1
     * @since 1.0.0
     */
    public void setNumberOfTouchesRequired(int value) {
        mTouchesRequired = value;
    }

    @SuppressWarnings ("checkstyle:cyclomaticcomplexity")
    @Override
    protected boolean onTouchEvent(final MotionEvent ev) {
        if (!isEnabled()) {
            return false;
        }

        final int action = ev.getAction();
        final int count = ev.getPointerCount();

        final boolean pointerUp = action == MotionEvent.ACTION_POINTER_UP;
        final int skipIndex = pointerUp ? ev.getActionIndex() : -1;

        // Determine focal point
        float sumX = 0, sumY = 0;
        for (int i = 0; i < count; i++) {
            if (skipIndex == i) {
                continue;
            }
            sumX += ev.getX(i);
            sumY += ev.getY(i);
        }

        final int div = pointerUp ? count - 1 : count;
        final float focusX = sumX / div;
        final float focusY = sumY / div;
        mCurrentLocation.set(focusX, focusY);

        switch (action & MotionEvent.ACTION_MASK) {
            case MotionEvent.ACTION_DOWN:
                removeMessages();

                mAlwaysInTapRegion = true;
                mNumTouches = count;
                mBegan = false;

                if (!mStarted) {
                    mFireEvents = false;
                    stopListenForOtherStateChanges();
                    setState(State.Possible);
                    mNumTaps = 0;
                    mStarted = true;
                } else {
                    mNumTaps++;
                }

                if (mNumTaps == mTapsRequired) {
                    mHandler.sendEmptyMessageAtTime(MESSAGE_LONG_PRESS, ev.getDownTime() + longPressTimeOut);
                } else {
                    long timeout = LONG_PRESS_TIMEOUT;
                    if (timeout >= longPressTimeOut) {
                        timeout = longPressTimeOut - 1;
                    }
                    mHandler.sendEmptyMessageDelayed(MESSAGE_FAILED, timeout);
                }

                mDownFocusX = focusX;
                mDownFocusY = focusY;
                break;

            case MotionEvent.ACTION_POINTER_DOWN:
                if (getState() == State.Possible && mStarted) {
                    removeMessages(MESSAGE_POINTER_UP);
                    mNumTouches = count;

                    if (mNumTouches > 1) {
                        if (mNumTouches > mTouchesRequired) {
                            removeMessages();
                            setState(State.Failed);
                        }
                    }

                    mDownFocusX = focusX;
                    mDownFocusY = focusY;
                } else if (inState(State.Began, State.Changed) && mStarted) {
                    mNumTouches = count;
                }
                break;

            case MotionEvent.ACTION_POINTER_UP:
                if (getState() == State.Possible && mStarted) {
                    removeMessages(MESSAGE_POINTER_UP);

                    mDownFocusX = focusX;
                    mDownFocusY = focusY;

                    Message message = mHandler.obtainMessage(MESSAGE_POINTER_UP);
                    message.arg1 = mNumTouches - 1;
                    mHandler.sendMessageDelayed(message, TAP_TIMEOUT);
                } else if (inState(State.Began, State.Changed)) {
                    if (mNumTouches - 1 < mTouchesRequired) {
                        setState(State.Ended);
                        if (mFireEvents) {
                            fireActionEvent();
                        }
                    }
                }
                break;

            case MotionEvent.ACTION_MOVE:
                if (getState() == State.Possible && mStarted) {

                    if (mAlwaysInTapRegion) {
                        final float deltaX = (focusX - mDownFocusX);
                        final float deltaY = (focusY - mDownFocusY);
                        final float distance = (deltaX * deltaX) + (deltaY * deltaY);

                        if (distance > mAllowableMovementSquare) {
                            logMessage(Log.WARN, "moved too much!: " + distance);
                            mAlwaysInTapRegion = false;
                            removeMessages();
                            setState(State.Failed);
                        }
                    }
                } else if (getState() == State.Began) {
                    if (!mBegan) {
                        final float deltaX = (focusX - mDownFocusX);
                        final float deltaY = (focusY - mDownFocusY);
                        final float distance = (deltaX * deltaX) + (deltaY * deltaY);

                        if (distance > mTouchSlopSquare) {
                            mBegan = true;

                            if (mFireEvents) {
                                setState(State.Changed);
                                fireActionEvent();
                            }
                        }
                    }
                } else if (getState() == State.Changed) {
                    setState(State.Changed);
                    if (mFireEvents) {
                        fireActionEvent();
                    }
                }
                break;

            case MotionEvent.ACTION_UP:
                removeMessages(MESSAGE_RESET, MESSAGE_POINTER_UP, MESSAGE_LONG_PRESS);

                if (getState() == State.Possible && mStarted) {
                    if (mNumTouches != mTouchesRequired) {
                        mStarted = false;
                        removeMessages();
                        setState(State.Failed);
                        postReset();
                    } else {
                        if (mNumTaps < mTapsRequired) {
                            removeMessages(MESSAGE_FAILED);
                            delayedFail();
                        } else {
                            mNumTaps = 0;
                            mStarted = false;
                            removeMessages();
                            setState(State.Failed);
                        }
                    }
                } else if (inState(State.Began, State.Changed)) {
                    mNumTaps = 0;
                    mStarted = false;
                    setState(State.Ended);
                    if (mFireEvents) {
                        fireActionEvent();
                    }
                } else {
                    mStarted = false;
                }
                break;

            case MotionEvent.ACTION_CANCEL:
                removeMessages();
                mStarted = false;
                mNumTaps = 0;
                setState(State.Cancelled);
                postReset();
                break;

            default:
                break;

        }

        return getCancelsTouchesInView();
    }

    private void removeMessages() {
        removeMessages(MESSAGE_FAILED, MESSAGE_RESET, MESSAGE_POINTER_UP, MESSAGE_LONG_PRESS);
    }

    private void postReset() {
        mHandler.sendEmptyMessage(MESSAGE_RESET);
    }

    private void delayedFail() {
        mHandler.sendEmptyMessageDelayed(MESSAGE_FAILED, DOUBLE_TAP_TIMEOUT);
    }

    private void handleFailed() {
        mStarted = false;
        mFireEvents = false;
        removeMessages();
        setState(State.Failed);
    }

    private void handleReset() {
        setState(State.Possible);
        mStarted = false;
        mFireEvents = false;
    }

    private void handleLongPress() {
        logMessage(Log.INFO, "handleLongPress");

        removeMessages(MESSAGE_FAILED);

        if (getState() == State.Possible && mStarted) {
            if (mNumTouches == mTouchesRequired && getDelegate().shouldBegin(this)) {
                setState(State.Began);
                if (null == getRequireFailureOf()) {
                    mFireEvents = true;
                    fireActionEvent();
                } else {
                    if (getRequireFailureOf().getState() == State.Failed) {
                        mFireEvents = true;
                        fireActionEvent();
                    } else {
                        listenForOtherStateChanges();
                        mFireEvents = false;
                        logMessage(Log.DEBUG, "waiting...");
                    }
                }
            } else {
                setState(State.Failed);
                mStarted = false;
                mNumTaps = 0;
            }
        }
    }

    @Override
    public float getCurrentLocationX() {
        return mCurrentLocation.x;
    }

    @Override
    public float getCurrentLocationY() {
        return mCurrentLocation.y;
    }

    /**
     * @param value The maximum movement of the fingers on the view before the gesture fails.
     * @since 1.0.0
     */
    public void setAllowableMovement(final float value) {
        mAllowableMovementSquare = value * value;
    }

    /**
     * @return The maximum allowed movement of the fingers on the view before the gesture fails.
     * @since 1.0.0
     */
    public float getAllowableMovement() {
        return mAllowableMovementSquare;
    }

    /**
     * @param value The minimum period fingers must press on the view for the gesture to be recognized.<br />
     *              Value is in milliseconds
     * @since 1.0.0
     */
    public void setMinimumPressDuration(long value) {
        longPressTimeOut = value;
    }

    /**
     * @return The minimum period fingers must press on the view for the gesture to be recognized.
     * @since 1.0.0
     */
    public long getMinimumPressDuration() {
        return longPressTimeOut;
    }

}

File: uigesturerecognizer/src/main/java/it/sephiroth/android/library/uigestures/UILongPressGestureRecognizer.kt
package it.sephiroth.android.library.uigestures

import android.content.Context
import android.graphics.PointF
import android.os.Message
import android.util.Log
import android.view.MotionEvent
import android.view.ViewConfiguration
import kotlin.math.max

/**
 * UILongPressGestureRecognizer looks for long-press gestures. The user must
 * press one or more fingers on a view and hold them there for a minimum period of time before the action triggers. While down,
 * the user’s fingers may not move more than a specified distance; if they move beyond the specified distance, the gesture fails.
 *
 * @author alessandro crugnola
 * @see [
 * https://developer.apple.com/reference/uikit/uilongpressgesturerecognizer](https://developer.apple.com/reference/uikit/uilongpressgesturerecognizer)
 */
@Suppress("MemberVisibilityCanBePrivate")
open class UILongPressGestureRecognizer(context: Context) : UIGestureRecognizer(context), UIContinuousRecognizer {

    /**
     * Maximum allowed timeout between consecutive taps (when tapsRequired > 1)
     */
    var doubleTapTimeout = DOUBLE_TAP_TIMEOUT
        set(value) {
            field = value

            if (value > longPressTimeout) {
                longPressTimeout = value + TIMEOUT_DELAY_MILLIS
            }
        }

    override var numberOfTouches: Int = 0
        internal set

    /**
     * The minimum period fingers must press on the view for the gesture to be recognized.<br></br>
     * Value is in milliseconds
     * @since 1.0.0
     */
    var longPressTimeout = max(LONG_PRESS_TIMEOUT, doubleTapTimeout)

    /**
     * The number of required touches for this recognizer to succeed.<br></br>
     * Default value is 1
     * @since 1.0.0
     */
    var touchesRequired = 1

    /**
     * The number of required taps for this recognizer to succeed.<br></br>
     * Default value is 1
     * @since 1.0.0
     */
    var tapsRequired = 1


    private var mAlwaysInTapRegion: Boolean = false
    private var mStarted: Boolean = false
    private val mStartLocation = PointF()
    private val mDownFocusLocation = PointF()
    private var mNumTaps = 1
    private var mBegan: Boolean = false

    val startLocationX: Float
        get() = mStartLocation.x

    val startLocationY: Float
        get() = mStartLocation.y

    /**
     * @return The maximum allowed movement of the fingers on the view before the gesture fails.
     * @since 1.2.4
     */
    var allowableMovement: Int

    /**
     * Minimum movement before we start collecting movements
     * @since 1.2.5
     */
    var scaledTouchSlop: Int

    /**
     * the duration in milliseconds between the first tap's up event and the second tap's down
     * event for an interaction to be considered a double-tap
     * @since 1.2.5
     */
    var scaledDoubleTapSlop: Int

    init {
        mStarted = false
        mBegan = false

        val configuration = ViewConfiguration.get(context)

        scaledTouchSlop = configuration.scaledTouchSlop
        scaledDoubleTapSlop = configuration.scaledDoubleTapSlop
        allowableMovement = configuration.scaledTouchSlop

        if (logEnabled) {
            logMessage(Log.INFO, "allowableMovement: $allowableMovement")
            logMessage(Log.INFO, "scaledTouchSlop: $scaledTouchSlop")
            logMessage(Log.INFO, "scaledDoubleTapSlop: $scaledDoubleTapSlop")
            logMessage(Log.INFO, "longPressTimeout: $longPressTimeout")
            logMessage(Log.INFO, "doubleTapTimeout: $doubleTapTimeout")
        }
    }

    override fun handleMessage(msg: Message) {
        when (msg.what) {
            MESSAGE_RESET -> {
                logMessage(Log.INFO, "handleMessage(MESSAGE_RESET)")
                handleReset()
            }

            MESSAGE_FAILED -> {
                logMessage(Log.INFO, "handleMessage(MESSAGE_FAILED)")
                handleFailed()
            }

            MESSAGE_POINTER_UP -> {
                logMessage(Log.INFO, "handleMessage(MESSAGE_POINTER_UP)")
                numberOfTouches = msg.arg1
            }

            MESSAGE_LONG_PRESS -> {
                logMessage(Log.INFO, "handleMessage(MESSAGE_LONG_PRESS)")
                handleLongPress()
            }

            else -> {
            }
        }
    }

    override fun onStateChanged(recognizer: UIGestureRecognizer) {
        logMessage(Log.VERBOSE, "onStateChanged(${recognizer.state?.name}, started: $mStarted)")

        if (recognizer.state === State.Failed && state === State.Began) {
            stopListenForOtherStateChanges()
            fireActionEventIfCanRecognizeSimultaneously()

            if (mBegan && hasBeganFiringEvents()) {
                state = State.Changed
            }

        } else if (recognizer.inState(State.Began, State.Ended) && mStarted && inState(State.Possible, State.Began)) {
            stopListenForOtherStateChanges()
            removeMessages()
            state = State.Failed
            setBeginFiringEvents(false)
            mStarted = false
        }
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        super.onTouchEvent(event)

        if (!isEnabled) {
            return false
        }

        val action = event.actionMasked
        val count = event.pointerCount

        when (action and MotionEvent.ACTION_MASK) {
            MotionEvent.ACTION_DOWN -> {
                if (!mStarted && !delegate?.shouldReceiveTouch?.invoke(this)!!) {
                    return cancelsTouchesInView
                }

                removeMessages()

                mAlwaysInTapRegion = true
                numberOfTouches = count
                mBegan = false

                if (!mStarted) {
                    stopListenForOtherStateChanges()
                    state = State.Possible
                    setBeginFiringEvents(false)
                    mNumTaps = 1
                    mStarted = true
                } else {
                    mNumTaps++

                    // if second tap is too far wawy from the first
                    val distance = mDownLocation.distance(mPreviousDownLocation)
                    logMessage(Log.VERBOSE, "distance: $distance")
                    if (distance > scaledDoubleTapSlop) {
                        logMessage(Log.WARN, "second touch too far away ($distance > $scaledDoubleTapSlop)")
                        handleFailed()
                        return cancelsTouchesInView
                    }
                }

                logMessage(Log.VERBOSE, "num taps: $mNumTaps, tapsRequired: $tapsRequired")

                if (mNumTaps == tapsRequired) {
                    mHandler.sendEmptyMessageAtTime(MESSAGE_LONG_PRESS, event.downTime + longPressTimeout)
                } else {
                    delayedFail()
                }

                mDownFocusLocation.set(mCurrentLocation)
                mStartLocation.set(mCurrentLocation)
            }

            MotionEvent.ACTION_POINTER_DOWN -> if (state == State.Possible && mStarted) {
                removeMessages(MESSAGE_POINTER_UP)
                numberOfTouches = count

                if (numberOfTouches > 1) {
                    if (numberOfTouches > touchesRequired) {
                        removeMessages()
                        state = State.Failed
                    }
                }

                mDownFocusLocation.set(mCurrentLocation)
                computeFocusPoint(event, mStartLocation)

            } else if (inState(State.Began, State.Changed) && mStarted) {
                numberOfTouches = count
            }

            MotionEvent.ACTION_POINTER_UP -> if (state == State.Possible && mStarted) {
                removeMessages(MESSAGE_POINTER_UP)

                mDownFocusLocation.set(mCurrentLocation)

                val message = mHandler.obtainMessage(MESSAGE_POINTER_UP)
                message.arg1 = numberOfTouches - 1
                mHandler.sendMessageDelayed(message, UIGestureRecognizer.TAP_TIMEOUT)

                computeFocusPoint(event, mStartLocation)


            } else if (inState(State.Began, State.Changed)) {
                if (numberOfTouches - 1 < touchesRequired) {
                    val began = hasBeganFiringEvents()
                    state = State.Ended

                    if (began) {
                        fireActionEvent()
                    }

                    setBeginFiringEvents(false)
                }
            }

            MotionEvent.ACTION_MOVE -> if (state == State.Possible && mStarted) {

                if (mAlwaysInTapRegion) {
                    val distance = mCurrentLocation.distance(mDownFocusLocation)
                    logMessage(Log.VERBOSE, "distance: $distance, allowableMovement: $allowableMovement")

                    if (distance > allowableMovement) {
                        logMessage(Log.WARN, "moved too much!: $distance > $allowableMovement")
                        mAlwaysInTapRegion = false
                        removeMessages()
                        state = State.Failed
                    }
                }
            } else if (state == State.Began) {
                if (!mBegan) {
                    val distance = mCurrentLocation.distance(mDownFocusLocation)
                    logMessage(Log.VERBOSE, "distance: $distance, allowableMovement: $scaledTouchSlop")

                    if (distance > scaledTouchSlop) {
                        mBegan = true

                        if (hasBeganFiringEvents()) {
                            state = State.Changed
                            fireActionEvent()
                        }
                    }
                }
            } else if (state == State.Changed) {
                state = State.Changed
                if (hasBeganFiringEvents()) {
                    fireActionEvent()
                }
            }

            MotionEvent.ACTION_UP -> {
                removeMessages(MESSAGE_RESET, MESSAGE_POINTER_UP, MESSAGE_LONG_PRESS)

                if (state == State.Possible && mStarted) {
                    if (numberOfTouches != touchesRequired) {
                        mStarted = false
                        removeMessages()
                        state = State.Failed
                        postReset()
                    } else {
                        if (mNumTaps < tapsRequired) {
                            removeMessages(MESSAGE_FAILED)
                            delayedFail()
                        } else {
                            mNumTaps = 1
                            mStarted = false
                            removeMessages()
                            state = State.Failed
                        }
                    }
                } else if (inState(State.Began, State.Changed)) {
                    mNumTaps = 1
                    mStarted = false
                    val began = hasBeganFiringEvents()
                    state = State.Ended
                    if (began) {
                        fireActionEvent()
                    }
                    postReset()
                } else {
                    mStarted = false
                    postReset()
                }
                setBeginFiringEvents(false)
            }

            MotionEvent.ACTION_CANCEL -> {
                removeMessages()
                mStarted = false
                mNumTaps = 1
                state = State.Cancelled
                postReset()
            }

            else -> {
            }
        }

        return cancelsTouchesInView
    }

    override fun removeMessages() {
        removeMessages(MESSAGE_FAILED, MESSAGE_RESET, MESSAGE_POINTER_UP, MESSAGE_LONG_PRESS)
    }

    private fun postReset() {
        mHandler.sendEmptyMessage(MESSAGE_RESET)
    }

    private fun delayedFail() {
        mHandler.sendEmptyMessageDelayed(MESSAGE_FAILED, doubleTapTimeout)
    }

    private fun handleFailed() {
        removeMessages()
        state = State.Failed
        setBeginFiringEvents(false)
        mStarted = false
    }

    private fun handleReset() {
        state = State.Possible
        mStarted = false
    }

    override fun reset() {
        super.reset()
        handleReset()
    }

    private fun handleLongPress() {
        logMessage(Log.INFO, "handleLongPress")

        removeMessages(MESSAGE_FAILED)

        if (state === State.Possible && mStarted) {
            if (numberOfTouches == touchesRequired && delegate?.shouldBegin?.invoke(this)!!) {
                state = State.Began
                if (null == requireFailureOf) {
                    fireActionEventIfCanRecognizeSimultaneously()
                } else {
                    when {
                        requireFailureOf!!.state === State.Failed -> fireActionEventIfCanRecognizeSimultaneously()
                        requireFailureOf!!.inState(State.Began, State.Changed, State.Ended) -> {
                            state = State.Failed
                            setBeginFiringEvents(false)
                            mStarted = false
                            mNumTaps = 1
                        }
                        else -> {
                            listenForOtherStateChanges()
                            setBeginFiringEvents(false)
                            logMessage(Log.DEBUG, "waiting...")
                        }
                    }
                }
            } else {
                state = State.Failed
                setBeginFiringEvents(false)
                mStarted = false
                mNumTaps = 1
            }
        }
    }

    private fun fireActionEventIfCanRecognizeSimultaneously() {
        if (inState(State.Changed, State.Ended)) {
            setBeginFiringEvents(true)
            fireActionEvent()
        } else {
            if (delegate!!.shouldRecognizeSimultaneouslyWithGestureRecognizer(this)) {
                setBeginFiringEvents(true)
                fireActionEvent()
            }
        }
    }

    override fun hasBeganFiringEvents(): Boolean {
        return super.hasBeganFiringEvents() && inState(State.Began, State.Changed)
    }

    companion object {

        // request to change the current state to Failed
        private const val MESSAGE_FAILED = 1
        // request to change the current state to Possible
        private const val MESSAGE_RESET = 2
        // we handle the action_pointer_up received in the onTouchEvent with a delay
        // in order to check how many fingers were actually down when we're checking them
        // in the action_up.
        private const val MESSAGE_POINTER_UP = 3
        // post handle the long press event
        private const val MESSAGE_LONG_PRESS = 4
    }

}
